File: api/services/pricerunner_client.py
"""
PriceRunner API client — fetches live product data for article generation.

WHY THIS EXISTS
  The article generator needs real product data (price, specs, images, merchant count)
  to write factual, trustworthy Danish affiliate articles. PriceRunner's public web API
  (no auth required) is our source. This is a direct Python port of the TypeScript
  reference implementation in archive/src/scraper/pricerunner-client.ts.

WHY NO OFFICIAL API
  The formal PriceRunner publisher API (via Rune) is still pending. This client hits the
  same endpoints the browser hits, with browser-like headers. It covers everything we need
  for product data and is already proven in the TS implementation.

WHY ASYNC
  The FastAPI app is async throughout. Using httpx.AsyncClient keeps everything on the
  same event loop — no thread pool overhead and no sync blocking.

KEY DECISIONS
  - Rate limit: 1 req/sec minimum. PriceRunner blocks faster bursts.
  - Backoff: exponential on 429/5xx, max 4 retries. Fail fast on client errors.
  - Cache: 24hr in-memory per category. Category data doesn't change intraday.
  - User-Agent rotation: reduces risk of UA-based blocks on repeated requests.
  - Product IDs prefixed with "pr_" to distinguish from other future data sources.
  - Fetching by URL: extract category_id + product_id from URL, fetch category,
    filter to the specific product. No confirmed direct-product endpoint exists in the
    public API (see archive/docs/pricerunner-api-reference.md). Size=100 covers nearly
    all cases since the user is linking to a specific product they know about.
"""

import asyncio
import logging
import math
import random
import re
import time
from dataclasses import dataclass, field
from typing import Optional

import httpx

logger = logging.getLogger(__name__)

# ─── User-Agent rotation ───────────────────────────────────────────────────────
_USER_AGENTS = [
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36",
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 14_4) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.4 Safari/605.1.15",
    "Mozilla/5.0 (X11; Linux x86_64; rv:125.0) Gecko/20100101 Firefox/125.0",
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:126.0) Gecko/20100101 Firefox/126.0",
]


def _random_ua() -> str:
    return random.choice(_USER_AGENTS)


# ─── Rate limiting (1 req/sec minimum) ────────────────────────────────────────
_last_request_at: float = 0.0
_rate_limit_lock = asyncio.Lock()


async def _rate_limit() -> None:
    global _last_request_at
    async with _rate_limit_lock:
        now = time.monotonic()
        elapsed = now - _last_request_at
        if elapsed < 1.0:
            await asyncio.sleep(1.0 - elapsed)
        _last_request_at = time.monotonic()


# ─── Exponential backoff ───────────────────────────────────────────────────────
async def _with_backoff(fn, max_retries: int = 4):
    """Retry fn() on 429/5xx with exponential backoff. Raises on client errors."""
    for attempt in range(max_retries + 1):
        try:
            return await fn()
        except httpx.HTTPStatusError as e:
            status = e.response.status_code
            retryable = status == 429 or status == 503 or status >= 500
            if not retryable or attempt == max_retries:
                raise
            delay = min(1.0 * (2 ** attempt) + random.random() * 0.5, 30.0)
            logger.warning(
                "PriceRunner HTTP %s — retry %d/%d in %.1fs",
                status, attempt + 1, max_retries, delay,
            )
            await asyncio.sleep(delay)
    raise RuntimeError("unreachable")


# ─── 24-hour in-memory cache ───────────────────────────────────────────────────
_CACHE_TTL = 24 * 60 * 60  # seconds

@dataclass
class _CacheEntry:
    data: object
    expires_at: float


_cache: dict[str, _CacheEntry] = {}


def _cache_get(key: str):
    entry = _cache.get(key)
    if entry is None or time.monotonic() > entry.expires_at:
        _cache.pop(key, None)
        return None
    return entry.data


def _cache_set(key: str, data: object) -> None:
    _cache[key] = _CacheEntry(data=data, expires_at=time.monotonic() + _CACHE_TTL)


# ─── Base URL per country ──────────────────────────────────────────────────────
_PR_BASE: dict[str, str] = {
    "DK": "https://www.pricerunner.dk",
    "SE": "https://www.pricerunner.se",
    "GB": "https://www.pricerunner.com",
}


def _base(country: str) -> str:
    return _PR_BASE.get(country.upper(), "https://www.pricerunner.dk")


# ─── Internal product model ────────────────────────────────────────────────────
@dataclass
class RawProduct:
    """
    Normalised product data ready for the brief builder.
    Mirrors the TypeScript RawProduct type in archive/src/services/product-store.ts.
    All fields are required — brief builder will reject incomplete products.
    """
    id: str             # "pr_{pricerunner_id}", e.g. "pr_3332774746"
    name: str
    category: str       # internal slug, e.g. "kaffemaskiner"
    price_kr: float
    retailer: str       # cheapest merchant name
    affiliate_url: str  # full pricerunner.dk product URL
    image_url: str
    popularity_score: float
    out_of_stock: bool
    specs: dict[str, str] = field(default_factory=dict)
    # specs may contain: brand, rating, watchedLabel, priceDrop, popularityRank, merchantCount


# ─── Category ID → internal slug map ──────────────────────────────────────────
# Matches CATEGORY_ID_MAP in archive/src/scraper/pricerunner-client.ts
_CATEGORY_SLUG: dict[str, str] = {
    "81":   "frituregryder-airfryere",
    "82":   "kaffemaskiner",
    "250":  "ismaskiner",
    "14":   "vaskemaskiner",
    "1613": "robotstoevsugere",
    "1595": "robotplaeneklippere",
    "335":  "grill",
    "120":  "havemaskiner",
    "638":  "hojtryks-hedvandsrensere",
    "345":  "elvaerktoej",
    "1258": "bore-skruemaskiner",
    "1260": "elsave",
}


# ─── Popularity scoring ────────────────────────────────────────────────────────
def _compute_popularity_score(p: dict) -> float:
    """
    Score a product by demand signals. Higher = more popular.
    Mirrors computePopularityScore() in archive/src/scraper/pricerunner-client.ts.

    Scoring:
      watchers  → 200+ = 40pts, 100+ = 30pts, 50+ = 20pts, any WATCHED = 10pts
      rank      → 1=30pts, 2=25pts, 3=20pts, top10=10pts
      rating    → ≥4.5 × ln(reviews+1), max 20pts
      merchants → >20=5pts, >10=3pts, >5=1pt
    """
    score = 0.0

    ribbon = p.get("ribbon") or {}
    ribbon_val = ribbon.get("value", "")
    try:
        watch_num = int(ribbon_val)
        if watch_num >= 200:
            score += 40
        elif watch_num >= 100:
            score += 30
        elif watch_num >= 50:
            score += 20
        else:
            score += 10
    except (ValueError, TypeError):
        if ribbon.get("type") == "WATCHED":
            score += 10

    rank_data = p.get("rank") or {}
    rank = rank_data.get("rank")
    if rank is not None:
        if rank == 1:
            score += 30
        elif rank == 2:
            score += 25
        elif rank == 3:
            score += 20
        elif rank <= 10:
            score += 10

    rating = p.get("rating") or {}
    avg = rating.get("average") or 0
    cnt = rating.get("count") or 0
    if avg >= 4.5 and cnt > 0:
        score += min(20, round(avg * math.log(cnt + 1)))

    merchants = (p.get("previewMerchants") or {}).get("count") or 0
    if merchants > 20:
        score += 5
    elif merchants > 10:
        score += 3
    elif merchants > 5:
        score += 1

    return score


# ─── V4 API response → RawProduct ─────────────────────────────────────────────
def _map_v4_product(p: dict, base_url: str, category_id: str) -> RawProduct:
    """Map a raw PriceRunner v4 API product dict to our internal RawProduct model."""
    def absolute(url: Optional[str]) -> str:
        if not url:
            return base_url
        return url if url.startswith("http") else f"{base_url}{url}"

    price_raw = (p.get("lowestPrice") or {}).get("amount")
    price_kr = (
        float(price_raw)
        if price_raw is not None
        else float((p.get("cheapestOffer") or {}).get("price", {}).get("amount", 0) or 0)
    )

    retailer = (
        (p.get("cheapestOffer") or {}).get("merchant", {}).get("name")
        or ((p.get("topOffers") or [{}])[0]).get("merchant", {}).get("name")
        or "PriceRunner"
    )

    image_data = p.get("image") or {}
    image_url = absolute(image_data.get("url") or image_data.get("path"))
    affiliate_url = absolute(p.get("url"))

    internal_category = _CATEGORY_SLUG.get(category_id, category_id)

    specs: dict[str, str] = {}
    brand = (p.get("brand") or {}).get("name")
    if brand:
        specs["brand"] = brand
    desc = p.get("description")
    if desc:
        specs["description"] = desc
    rating = p.get("rating") or {}
    if rating.get("average") is not None:
        specs["rating"] = f"{rating['average']} ({rating.get('count', 0)} reviews)"
    ribbon = p.get("ribbon") or {}
    if ribbon.get("type"):
        specs["ribbon"] = ribbon["type"]
    if ribbon.get("type") == "WATCHED" and ribbon.get("value"):
        num = abs(float(ribbon["value"]))
        specs["watchedLabel"] = f"{round(num)}+"
    price_drop = p.get("priceDrop") or {}
    if price_drop.get("percent") is not None:
        specs["priceDrop"] = f"{price_drop['percent']}%"
    rank_data = p.get("rank") or {}
    if rank_data.get("rank") is not None:
        specs["popularityRank"] = str(rank_data["rank"])
    merchants = (p.get("previewMerchants") or {}).get("count")
    if merchants is not None:
        specs["merchantCount"] = str(merchants)

    popularity = _compute_popularity_score(p)

    return RawProduct(
        id=f"pr_{p['id']}",
        name=p.get("name", ""),
        category=internal_category,
        price_kr=price_kr,
        retailer=retailer,
        affiliate_url=affiliate_url,
        image_url=image_url,
        popularity_score=popularity,
        out_of_stock=p.get("outOfStock", False),
        specs=specs,
    )


# ─── Public API ────────────────────────────────────────────────────────────────

def extract_product_info_from_url(url: str) -> Optional[tuple[str, str]]:
    """
    Extract (category_id, product_id) from a PriceRunner product listing URL.

    PriceRunner URL format: /pl/{categoryId}-{productId}/...
    Example: https://www.pricerunner.dk/pl/82-3332774746/Kaffemaskiner/Ninja-...
      → category_id="82", product_id="3332774746"

    Returns None if the URL doesn't match the expected pattern.
    """
    match = re.search(r"/pl/(\d+)-(\d+)", url)
    if not match:
        logger.warning("Could not extract product info from URL: %s", url)
        return None
    return match.group(1), match.group(2)


async def fetch_products_by_category(
    category_id: str,
    country: str = "DK",
    size: int = 30,
) -> list[RawProduct]:
    """
    Fetch products for a PriceRunner category, sorted by popularity.

    Uses the v4 category browse endpoint documented in
    archive/docs/pricerunner-api-reference.md. Results are cached for 24 hours
    since category data doesn't change intraday.
    """
    cache_key = f"category:{category_id}:{country}:{size}"
    cached = _cache_get(cache_key)
    if cached is not None:
        return cached

    await _rate_limit()
    base = _base(country)
    country_upper = country.upper()
    country_lower = country.lower()

    async def _fetch():
        async with httpx.AsyncClient(timeout=15.0) as client:
            resp = await client.get(
                f"{base}/{country_lower}/api/search-edge-rest/public/search/category/v4/{country_upper}/{category_id}",
                params={"size": size, "sorting": "POPULARITY", "device": "desktop"},
                headers={"User-Agent": _random_ua(), "Accept": "application/json"},
            )
            resp.raise_for_status()
            data = resp.json()
            return data.get("products") or data.get("results") or []

    raw_products = await _with_backoff(_fetch)
    products = [_map_v4_product(p, base, category_id) for p in raw_products]
    _cache_set(cache_key, products)
    return products


async def fetch_product_from_url(url: str, country: str = "DK") -> Optional[RawProduct]:
    """
    Fetch a single product given its PriceRunner listing URL.

    Strategy:
      1. Extract category_id and product_id from the URL path
      2. Fetch up to 100 products from that category (sorted by popularity)
      3. Filter to the specific product by ID

    Why category fetch instead of a direct product endpoint:
      The public API only has a confirmed category browse endpoint. A size=100 fetch
      covers virtually all products a user would link to — if they're linking it,
      it's likely notable enough to appear in the top 100. If it doesn't appear
      (e.g. very niche or newly listed), we log a warning and return None.
    """
    info = extract_product_info_from_url(url)
    if not info:
        return None

    category_id, product_id = info
    target_id = f"pr_{product_id}"

    # Fetch up to 100 to maximise chance of finding the specific product
    products = await fetch_products_by_category(category_id, country=country, size=100)

    found = next((p for p in products if p.id == target_id), None)
    if not found:
        logger.warning(
            "Product %s not found in top 100 of category %s. "
            "It may be outside the popularity window.",
            product_id, category_id,
        )
    return found

File: archive/src/scraper/pricerunner-client.ts
import axios from "axios";
import type { RawProduct } from "../services/product-store.js";

// ─── User-Agent rotation ─────────────────────────────────────────────────────
const USER_AGENTS = [
  "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36",
  "Mozilla/5.0 (Macintosh; Intel Mac OS X 14_4) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.4 Safari/605.1.15",
  "Mozilla/5.0 (X11; Linux x86_64; rv:125.0) Gecko/20100101 Firefox/125.0",
  "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:126.0) Gecko/20100101 Firefox/126.0",
];

function randomUA(): string {
  return USER_AGENTS[Math.floor(Math.random() * USER_AGENTS.length)]!;
}

// ─── Exponential backoff ─────────────────────────────────────────────────────
export async function withBackoff<T>(fn: () => Promise<T>, maxRetries = 4): Promise<T> {
  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    try {
      return await fn();
    } catch (err: unknown) {
      const status = axios.isAxiosError(err) ? err.response?.status : undefined;
      const retryable = status === 429 || status === 503 || (status !== undefined && status >= 500);
      if (!retryable || attempt === maxRetries) throw err;
      const delayMs = Math.min(1000 * 2 ** attempt + Math.random() * 500, 30_000);
      console.warn(`[pricerunner] HTTP ${status} - retry ${attempt + 1}/${maxRetries} in ${Math.round(delayMs)}ms`);
      await new Promise((r) => setTimeout(r, delayMs));
    }
  }
  throw new Error("unreachable");
}

// ─── PriceRunner base URLs per country ───────────────────────────────────────
const PR_BASE_BY_COUNTRY: Record<string, string> = {
  DK: "https://www.pricerunner.dk",
  SE: "https://www.pricerunner.se",
  GB: "https://www.pricerunner.com",
};

function getBase(country: string): string {
  return PR_BASE_BY_COUNTRY[country.toUpperCase()] ?? "https://www.pricerunner.dk";
}

// ─── Rate limiting (1000ms minimum between requests) ─────────────────────────
let _lastRequestAt = 0;
async function rateLimit(): Promise<void> {
  const now = Date.now();
  const elapsed = now - _lastRequestAt;
  if (elapsed < 1000) await new Promise((r) => setTimeout(r, 1000 - elapsed));
  _lastRequestAt = Date.now();
}

// ─── 24-hour in-memory cache ──────────────────────────────────────────────────
interface CacheEntry<T> { data: T; expiresAt: number }
const _cache = new Map<string, CacheEntry<unknown>>();
const CACHE_TTL_MS = 24 * 60 * 60 * 1000;

function cacheGet<T>(key: string): T | undefined {
  const entry = _cache.get(key);
  if (!entry || Date.now() > entry.expiresAt) { _cache.delete(key); return undefined; }
  return entry.data as T;
}
function cacheSet<T>(key: string, data: T): void {
  _cache.set(key, { data, expiresAt: Date.now() + CACHE_TTL_MS });
}

// ─── v4 API response types ────────────────────────────────────────────────────
export interface V4Product {
  id: string;
  name: string;
  lowestPrice?: { amount: string; currency: string };
  cheapestOffer?: { price?: { amount: number }; merchant?: { name: string } };
  image?: { url?: string; path?: string };
  url?: string;
  brand?: { name: string };
  description?: string;
  rating?: { average: number; count: number };
  ribbon?: { type: string; value?: string };
  priceDrop?: { percent: number };
  topOffers?: Array<{ merchant?: { name: string } }>;
  rank?: { rank: number };
  previewMerchants?: { count: number };
  outOfStock?: boolean;
}

interface V4CategoryResponse {
  products?: V4Product[];
  // category browse endpoint wraps in results
  results?: V4Product[];
}

interface V4SuggestHit {
  id?: string;
  name?: string;
  type?: string;   // "CATEGORY" | "BRAND" | "PRODUCT" etc.
  url?: string;
}

interface V4SuggestResponse {
  products?: V4Product[];       // usually empty for keyword queries
  suggestions?: V4SuggestHit[]; // category/brand autocomplete hints
}

// ─── Maps PriceRunner category ID → our internal category name ───────────────
export const CATEGORY_ID_MAP: Record<string, string> = {
  "27": "laptops",
  "94": "headphones",
  "1": "phones",
  "2": "tvs",
  // Hjem & Husholdning
  "81": "frituregryder-airfryere",
  "82": "kaffemaskiner",
  "250": "ismaskiner",
  "14": "vaskemaskiner",
  "1613": "robotstoevsugere",
  // Have & Udemiljø
  "1595": "robotplaeneklippere",
  "335": "grill",
  "120": "havemaskiner",
  "638": "hojtryks-hedvandsrensere",
  // Værktøj
  "345": "elvaerktoej",
  "1258": "bore-skruemaskiner",
  "1260": "elsave",
};

// Legacy slug map kept for seed script compatibility
export const CATEGORY_MAP: Record<string, string> = {
  "baerbare-computere-17": "laptops",
  "hoeretelefoner-2451": "headphones",
  "mobiltelefoner-1": "phones",
  "tv-1344": "tvs",
};

function makeAbsoluteUrl(url: string | undefined, base: string): string {
  if (!url) return base;
  if (url.startsWith("http")) return url;
  return `${base}${url}`;
}

/** Compute a popularity score for ranking products within a category.
 *  Higher = more demand signal. Used by product-store to pick the best products.
 *
 *  Scoring:
 *    watchedCount  →  200+ = 40pts, 100+ = 30pts, 50+ = 20pts, any ribbon = 10pts
 *    rank.rank     →  1 = 30pts, 2 = 25pts, 3 = 20pts, top10 = 10pts, else 0
 *    rating        →  ≥4.5 × ln(reviews+1) bonus (max 20pts)
 *    merchantCount →  >20 = 5pts, >10 = 3pts, >5 = 1pt
 */
export function computePopularityScore(p: V4Product): number {
  let score = 0;

  // Watcher ribbon
  const ribbonVal = p.ribbon?.value ?? "";
  const watchNum = parseInt(ribbonVal, 10);
  if (!isNaN(watchNum)) {
    if (watchNum >= 200) score += 40;
    else if (watchNum >= 100) score += 30;
    else if (watchNum >= 50) score += 20;
    else score += 10;
  } else if (p.ribbon?.type === "WATCHED") {
    score += 10;
  }

  // Popularity rank
  const rank = p.rank?.rank;
  if (rank !== undefined) {
    if (rank === 1) score += 30;
    else if (rank === 2) score += 25;
    else if (rank === 3) score += 20;
    else if (rank <= 10) score += 10;
  }

  // Rating quality × log(volume)
  const avg = p.rating?.average ?? 0;
  const cnt = p.rating?.count ?? 0;
  if (avg >= 4.5 && cnt > 0) {
    score += Math.min(20, Math.round(avg * Math.log(cnt + 1)));
  }

  // Merchant depth
  const merchants = p.previewMerchants?.count ?? 0;
  if (merchants > 20) score += 5;
  else if (merchants > 10) score += 3;
  else if (merchants > 5) score += 1;

  return score;
}

export function mapV4Product(p: V4Product, base: string, internalCategory: string): RawProduct {
  const priceRaw = p.lowestPrice?.amount;
  const priceKr = priceRaw !== undefined
    ? parseFloat(priceRaw)
    : (p.cheapestOffer?.price?.amount ?? 0);

  const imageUrl = makeAbsoluteUrl(p.image?.url ?? p.image?.path, base);
  const affiliateUrl = makeAbsoluteUrl(p.url, base);

  // Retailer: prefer cheapestOffer merchant (same price we display), fall back to topOffers
  const retailer =
    p.cheapestOffer?.merchant?.name ??
    p.topOffers?.[0]?.merchant?.name ??
    "PriceRunner";

  const specs: Record<string, string> = {};
  if (p.brand?.name) specs["brand"] = p.brand.name;
  if (p.description) specs["description"] = p.description;
  if (p.rating?.average !== undefined) specs["rating"] = `${p.rating.average} (${p.rating.count ?? 0} reviews)`;
  if (p.ribbon?.type) specs["ribbon"] = p.ribbon.type;
  if (p.ribbon?.type === "WATCHED" && p.ribbon?.value) {
    // Format as integer count with "+" suffix, strip any minus sign
    const num = Math.abs(parseFloat(p.ribbon.value));
    specs["watchedLabel"] = `${Math.round(num)}+`;
  }
  if (p.priceDrop?.percent !== undefined) specs["priceDrop"] = `${p.priceDrop.percent}%`;
  if (p.rank?.rank !== undefined) specs["popularityRank"] = String(p.rank.rank);
  if (p.previewMerchants?.count !== undefined) specs["merchantCount"] = String(p.previewMerchants.count);

  const popularityScore = computePopularityScore(p);
  if (popularityScore > 0) specs["popularityScore"] = String(popularityScore);

  return {
    id: `pr_${p.id}`,
    name: p.name,
    category: internalCategory,
    priceKr,
    retailer,
    affiliateUrl,
    imageUrl,
    popularityScore,
    outOfStock: p.outOfStock ?? false,
    specs,
  };
}

// ─── Category Browse v4 ───────────────────────────────────────────────────────
export interface AfFilter {
  attributeId: string;
  valueId: string;
}

export async function fetchProductsByCategoryId(
  categoryId: string,
  country = "DK",
  size = 30,
  afFilters?: AfFilter[]
): Promise<RawProduct[]> {
  const filterKey = afFilters?.map(f => `${f.attributeId}=${f.valueId}`).join(':') ?? '';
  const cacheKey = `pricerunner-category:${categoryId}:${country}:${filterKey}`;
  const cached = cacheGet<RawProduct[]>(cacheKey);
  if (cached) return cached;

  await rateLimit();
  const base = getBase(country);
  const countryUpper = country.toUpperCase();
  const countryLower = country.toLowerCase();
  const internalCategory = CATEGORY_ID_MAP[categoryId] ?? categoryId;

  const results = await withBackoff(async () => {
    const params: Record<string, string | number> = { size, sorting: "POPULARITY", device: "desktop" };
    if (afFilters) {
      for (const f of afFilters) {
        params[`af_${f.attributeId}`] = f.valueId;
      }
    }
    const res = await axios.get<V4CategoryResponse>(
      `${base}/${countryLower}/api/search-edge-rest/public/search/category/v4/${countryUpper}/${categoryId}`,
      {
        params,
        headers: { "User-Agent": randomUA(), "Accept": "application/json" },
        timeout: 15_000,
      }
    );
    return res.data.products ?? res.data.results ?? [];
  });

  const mapped = results.map((p) => mapV4Product(p, base, internalCategory));
  cacheSet(cacheKey, mapped);
  return mapped;
}

// ─── Keyword Search (suggest endpoint) ───────────────────────────────────────
export async function searchProductsByKeyword(
  term: string,
  country = "DK"
): Promise<RawProduct[]> {
  await rateLimit();
  const base = getBase(country);
  const countryUpper = country.toUpperCase();
  const countryLower = country.toLowerCase();

  const results = await withBackoff(async () => {
    const res = await axios.get<V4SuggestResponse>(
      `${base}/${countryLower}/api/instant-search-edge-rest/public/search/suggest/${countryUpper}`,
      {
        params: { q: term },
        headers: { "User-Agent": randomUA(), "Accept": "application/json" },
        timeout: 10_000,
      }
    );
    return res.data.products ?? [];
  });

  return results.map((p) => mapV4Product(p, base, "unknown"));
}

// ─── Legacy: category slug → v4 by mapping category ID ───────────────────────
/** Used by the seed script. Maps old slug names to IDs and calls v4. */
export async function fetchProductsByCategory(
  categorySlug: string,
  limit = 10
): Promise<RawProduct[]> {
  // Extract numeric ID from slug (e.g. "baerbare-computere-17" → "17")
  const match = /(\d+)$/.exec(categorySlug);
  const categoryId = match?.[1];
  if (!categoryId) throw new Error(`Cannot extract category ID from slug: ${categorySlug}`);
  const products = await fetchProductsByCategoryId(categoryId, "DK", Math.max(limit, 30));
  return products.slice(0, limit);
}

// ─── Category Tree Discovery ─────────────────────────────────────────────────

interface MenuCategory {
  id: string;
  name: string;
  url?: string | null;
  children?: Array<{ id: string; name: string; url?: string | null }>;
}

interface MenuResponse {
  id: string;
  name: string;
  categories: MenuCategory[];
}

export interface DiscoveredLeafCategory {
  id: string;
  name: string;
  parentId: string;
  parentName: string;
}

/** Discover all leaf categories under a PriceRunner topic.
 *  Hits the navigation/menu endpoint once, caches to disk.
 *  Topic IDs look like "t34", "t1424" etc.
 */
export async function discoverLeafCategories(
  topicId: string,
  country = "DK"
): Promise<DiscoveredLeafCategory[]> {
  const cacheKey = `pricerunner-tree:${topicId}:${country}`;
  const cached = cacheGet<DiscoveredLeafCategory[]>(cacheKey);
  if (cached) return cached;

  await rateLimit();
  const base = getBase(country);
  const countryLower = country.toLowerCase();
  const countryUpper = country.toUpperCase();

  const res = await withBackoff(async () => {
    return axios.get<MenuResponse>(
      `${base}/${countryLower}/api/seo-edge-rest/public/navigation/menu/${countryUpper}/hierarchy/${topicId}`,
      {
        headers: { "User-Agent": randomUA(), "Accept": "application/json" },
        timeout: 15_000,
      }
    );
  });

  const data = res.data;
  const leaves: DiscoveredLeafCategory[] = [];

  for (const cat of data.categories ?? []) {
    if (cat.children && cat.children.length > 0) {
      for (const child of cat.children) {
        // Skip filter combinations (IDs with dashes like "100003649-100015017")
        if (/\d+-\d+/.test(child.id)) continue;
        leaves.push({
          id: child.id,
          name: child.name,
          parentId: cat.id,
          parentName: cat.name,
        });
      }
    }
  }

  cacheSet(cacheKey, leaves);
  return leaves;
}

/** Extract topic ID from a PriceRunner URL.
 *  Supports /t/{id}/Name and /cl/{id}/Name formats.
 */
export function extractTopicIdFromUrl(url: string): string | null {
  const tMatch = /\/t\/(\d+)\/[^\/]*$/.exec(url);
  if (tMatch) return `t${tMatch[1]}`;
  const clMatch = /\/cl\/(\d+)\/[^\/]*$/.exec(url);
  if (clMatch) return `cl${clMatch[1]}`;
  return null;
}
